queue:TestBreakerRecover is flakey or breaker is racey

<!--
Pro-tip: You can leave this block commented, and it still works!

Select the appropriate areas for your issue:
/area test-and-release

Classify what kind of issue this is:
/kind bug
/kind cleanup

-->

## Expected Behavior
In `knative/serving/pkg/queue` running the following command should never fail

```
go test -run TestBreakerRecover -count=1000000
```

## Actual Behavior

I've seen tests fail in [prow here](https://gubernator.knative.dev/build/knative-prow/pr-logs/pull/knative_serving/1983/pull-knative-serving-unit-tests/1038046832120303617/)  and locally
```
go test -run TestBreakerRecover -count=1000000 ./...
--- FAIL: TestBreakerRecover (0.00s)
	breaker_test.go:75: Wanted [true true false false true true]. Got [true false true false true true].
--- FAIL: TestBreakerRecover (0.00s)
	breaker_test.go:75: Wanted [true true false false true true]. Got [true true false true true true].
--- FAIL: TestBreakerRecover (0.00s)
	breaker_test.go:75: Wanted [true true false false true true]. Got [true true false true true true].
--- FAIL: TestBreakerRecover (0.00s)
	breaker_test.go:75: Wanted [true true false false true true]. Got [true true true false true true].
--- FAIL: TestBreakerRecover (0.00s)
	breaker_test.go:75: Wanted [true true false false true true]. Got [true true false true true true].
--- FAIL: TestBreakerRecover (0.00s)
	breaker_test.go:75: Wanted [true true false false true true]. Got [true true false true true true].
--- FAIL: TestBreakerRecover (0.00s)
	breaker_test.go:75: Wanted [true true false false true true]. Got [true true true false true true].
FAIL
FAIL	github.com/knative/serving/pkg/queue	26.061s
```
I've also seen the test deadlock.

If you send a SIGABRT signal to the test it'll exit and print the goroutines
```
SIGABRT: abort
PC=0x108dfbb m=0 sigcode=0

goroutine 0 [idle]:
runtime.mach_semaphore_wait(0xd03, 0xe, 0x2b0c828, 0x1ee3a20, 0x2b0c8c0, 0x1ee3520, 0x7ffeefbff5e8, 0x10881d3, 0xffffffffffffffff, 0x1f05a20, ...)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/sys_darwin_amd64.s:540 +0xb
runtime.semasleep1(0xffffffffffffffff, 0x1f05a20)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/os_darwin.go:438 +0x52
runtime.semasleep.func1()
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/os_darwin.go:457 +0x33
runtime.semasleep(0xffffffffffffffff, 0x7ffeefbff658)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/os_darwin.go:456 +0x44
runtime.notesleep(0x1ee3b68)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/lock_sema.go:167 +0xe9
runtime.stopm()
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/proc.go:1952 +0xe5
runtime.findrunnable(0xc420034f00, 0x0)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/proc.go:2415 +0x2c8
runtime.schedule()
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/proc.go:2541 +0x151
runtime.park_m(0xc420463980)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/proc.go:2604 +0x81
runtime.mcall(0x0)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/asm_amd64.s:351 +0x5b

goroutine 1 [chan receive]:
testing.(*T).Run(0xc4204b5a40, 0x1a45604, 0x12, 0x1a8d640, 0xc42014bc01)
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:825 +0x597
testing.runTests.func1(0xc4204b5a40)
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:1063 +0xa5
testing.tRunner(0xc4204b5a40, 0xc42014bdb0)
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:777 +0x16e
testing.runTests(0xc42000c260, 0x1edbe20, 0xb, 0xb, 0xc42025c000)
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:1061 +0x4e2
testing.(*M).Run(0xc42025c000, 0x0)
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:978 +0x2ce
main.main()
	_testmain.go:62 +0x22b

goroutine 19 [chan receive]:
github.com/knative/serving/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x1ee31a0)
	/Users/dprotasowski/go/src/github.com/knative/serving/vendor/github.com/golang/glog/glog.go:882 +0xac
created by github.com/knative/serving/vendor/github.com/golang/glog.init.0
	/Users/dprotasowski/go/src/github.com/knative/serving/vendor/github.com/golang/glog/glog.go:410 +0x231

goroutine 21 [select]:
github.com/knative/serving/vendor/go.opencensus.io/stats/view.(*worker).start(0xc4200b4e00)
	/Users/dprotasowski/go/src/github.com/knative/serving/vendor/go.opencensus.io/stats/view/worker.go:144 +0x18f
created by github.com/knative/serving/vendor/go.opencensus.io/stats/view.init.0
	/Users/dprotasowski/go/src/github.com/knative/serving/vendor/go.opencensus.io/stats/view/worker.go:29 +0x9b

goroutine 3826 [chan send]:
github.com/knative/serving/pkg/queue.(*Breaker).Maybe(0xc420401bc0, 0xc4202b7f98, 0x0)
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker.go:63 +0x9d
github.com/knative/serving/pkg/queue.(*Breaker).concurrentRequest.func1(0xc4204818e0, 0xc420401bc0, 0xc4204818d0, 0xc4204aff10)
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker_test.go:125 +0xab
created by github.com/knative/serving/pkg/queue.(*Breaker).concurrentRequest
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker_test.go:123 +0x121

goroutine 3783 [semacquire]:
sync.runtime_SemacquireMutex(0xc420439d34, 0x900000000)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc420439d30)
	/usr/local/Cellar/go/1.10.3/libexec/src/sync/mutex.go:134 +0x172
github.com/knative/serving/pkg/queue.(*Breaker).concurrentRequest.func1.1()
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker_test.go:126 +0x47
github.com/knative/serving/pkg/queue.(*Breaker).Maybe(0xc420401bc0, 0xc4203c3f98, 0x0)
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker.go:67 +0xcd
github.com/knative/serving/pkg/queue.(*Breaker).concurrentRequest.func1(0xc420439d40, 0xc420401bc0, 0xc420439d30, 0xc4204bf340)
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker_test.go:125 +0xab
created by github.com/knative/serving/pkg/queue.(*Breaker).concurrentRequest
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker_test.go:123 +0x121

goroutine 3823 [chan receive]:
github.com/knative/serving/pkg/queue.unlock(0xc4204818d0, 0xc4204aff10)
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker_test.go:156 +0x56
github.com/knative/serving/pkg/queue.unlockAll(0xc4203fedc0, 0x4, 0x4)
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker_test.go:163 +0x6e
github.com/knative/serving/pkg/queue.TestBreakerRecover(0xc4204b5b30)
	/Users/dprotasowski/go/src/github.com/knative/serving/pkg/queue/breaker_test.go:68 +0xfb
testing.tRunner(0xc4204b5b30, 0x1a8d640)
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:777 +0x16e
created by testing.(*T).Run
	/usr/local/Cellar/go/1.10.3/libexec/src/testing/testing.go:824 +0x565

rax    0xe
rbx    0x1ee3a20
rcx    0x7ffeefbff588
rdx    0x7ffeefbff608
rdi    0xd03
rsi    0x1
rbp    0x7ffeefbff5c0
rsp    0x7ffeefbff588
r8     0xc420463a08
r9     0x8a17508
r10    0x1ee3520
r11    0x246
r12    0xc42005fda0
r13    0x1ee3a20
r14    0xc420463980
r15    0xc4204818d0
rip    0x108dfbb
rflags 0x246
cs     0x7
fs     0x0
gs     0x0
FAIL	github.com/knative/serving/pkg/queue	127.773s
```

--- a/pkg/queue/breaker_test.go
+++ b/pkg/queue/breaker_test.go
@@ -18,9 +18,11 @@ package queue
 
 import (
 	"reflect"
-	"runtime"
 	"sync"
 	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
 )
 
 type request struct {
@@ -28,11 +30,18 @@ type request struct {
 	accepted chan bool
 }
 
+func (r *request) wait() {
+	ok := <-r.accepted
+	// Requeue for next usage
+	r.accepted <- ok
+}
+
 func TestBreakerOverload(t *testing.T) {
 	b := NewBreaker(1, 1)             // Breaker capacity = 2
 	want := []bool{true, true, false} // Only first two requests will be processed
 
 	locks := b.concurrentRequests(3)
+
 	unlockAll(locks)
 
 	got := accepted(locks)
@@ -107,17 +116,22 @@ func TestBreakerLargeCapacityRecover(t *testing.T) {
 }
 
 // Attempts to perform a concurrent request against the specified breaker.
+// Will wait for request to either be performed, enqueued or rejected.
 func (b *Breaker) concurrentRequest() request {
-
-	// There is a brief window between when capacity is released and
-	// when it becomes available to the next request.  We yield here
-	// to reduce the likelihood that we hit that edge case.  E.g.
-	// without yielding `go test ./pkg/queue/breaker.* -count 10000`
-	// will fail about 3 runs.
-	runtime.Gosched()
-
 	r := request{lock: &sync.Mutex{}, accepted: make(chan bool, 1)}
 	r.lock.Lock()
+
+	if len(b.activeRequests) < cap(b.activeRequests) {
+		// Expect request to be performed
+		defer waitForQueue(b.activeRequests, len(b.activeRequests)+1)
+	} else if len(b.pendingRequests) < cap(b.pendingRequests) {
+		// Expect request to be queued
+		defer waitForQueue(b.pendingRequests, len(b.pendingRequests)+1)
+	} else {
+		// Expect request to be rejected
+		defer r.wait()
+	}
+
 	var start sync.WaitGroup
 	start.Add(1)
 	go func() {
@@ -142,6 +156,15 @@ func (b *Breaker) concurrentRequests(n int) []request {
 	return requests
 }
 
+func waitForQueue(queue chan token, size int) {
+	err := wait.PollImmediate(1*time.Millisecond, 100*time.Millisecond, func() (bool, error) {
+		return len(queue) == size, nil
+	})
+	if err != nil {
+		panic("timed out waiting for queue")
+	}
+}
+
 func accepted(requests []request) []bool {
 	got := make([]bool, len(requests))
 	for i, r := range requests {
@@ -153,9 +176,7 @@ func accepted(requests []request) []bool {
 func unlock(req request) {
 	req.lock.Unlock()
 	// Verify that function has completed
-	ok := <-req.accepted
-	// Requeue for next usage
-	req.accepted <- ok
+	req.wait()
 }
 
 func unlockAll(requests []request) {

